library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ModResourcePackProvider.java

```diff
@@ -0,0 +1,81 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.resource.loader.impl;
+
+import java.util.function.Consumer;
+
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.resource.ResourcePackProfile;
+import net.minecraft.resource.ResourcePackProvider;
+import net.minecraft.resource.ResourcePackSource;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.text.Text;
+import net.minecraft.text.TranslatableText;
+import net.minecraft.util.Formatting;
+
+/**
+ * Represents a resource pack provider for built-in mods resource packs and low-priority virtual resource packs.
+ */
+@ApiStatus.Internal
+public final class ModResourcePackProvider implements ResourcePackProvider {
+	/**
+	 * Equivalent to {@link ResourcePackSource#PACK_SOURCE_BUILTIN} but allows a different reference equality.
+	 */
+	public static final ResourcePackSource PACK_SOURCE_MOD_BUILTIN = nameAndSource("pack.source.builtin");
+	public static final ModResourcePackProvider CLIENT_RESOURCE_PACK_PROVIDER = new ModResourcePackProvider(ResourceType.CLIENT_RESOURCES);
+	public static final ModResourcePackProvider SERVER_RESOURCE_PACK_PROVIDER = new ModResourcePackProvider(ResourceType.SERVER_DATA);
+
+	private final ResourceType type;
+	private final ResourcePackProfile.Factory factory;
+
+	public ModResourcePackProvider(ResourceType type) {
+		this.type = type;
+		this.factory = (name, text, bl, supplier, metadata, initialPosition, source) ->
+				new ResourcePackProfile(name, text, bl, supplier, metadata, type, initialPosition, source);
+	}
+
+	/**
+	 * Registers the resource packs.
+	 *
+	 * @param consumer The resource pack profile consumer.
+	 */
+	public void register(Consumer<ResourcePackProfile> consumer) {
+		this.register(consumer, this.factory);
+	}
+
+	@Override
+	public void register(Consumer<ResourcePackProfile> profileAdder, ResourcePackProfile.Factory factory) {
+		/*
+			Register order rule in this provider:
+			1. Mod built-in resource packs
+			Register order rule globally:
+			1. Default (with mod resource packs bundled) and Vanilla built-in resource packs
+			2. Mod built-in resource packs
+			3. Low priority virtual resource packs
+			4. User resource packs
+			5. (Invisible) High-priority virtual resource packs
+		 */
+
+		ResourceLoaderImpl.registerBuiltinResourcePacks(this.type, profileAdder, factory);
+	}
+
+	private static ResourcePackSource nameAndSource(String source) {
+		Text text = new TranslatableText(source);
+		return name -> (new TranslatableText("pack.nameAndSource", name, text)).formatted(Formatting.GRAY);
+	}
```

Comment:
Isn't this just a copy of `ResourcePackSource.nameAndSource`? Why not use that instead?

Reply:
Reference equality is different, which is used to our advantage.

Reply:
You mean like for the lambda?
Won't that be created every time anyway, considering it uses a variable from the outer scope (`text`)

Reply:
What I mean is the lambda is of the type `ResourcePackSource`, a `ResourcePackProfile` has a method to get the source which returns `ResourcePackSource`, we then use a reference equality to differentiate the quilt's one from the Vanilla's one.

Reply:
Well, yes, but can't you use the same method to make the quilt `ResourcePackSource`?

Reply:
It's package-private and I found the use of an accessor quite ridiculous in this case.

Reply:
> It's package-private

Is it though? I made line 39 use `ResourcePackSource`'s method and it built completely fine.

Reply:
I have been deceived.